src/games/establishment.c

```diff
@@ -54,42 +58,100 @@
  */
 int igraph_establishment_game(igraph_t *graph, igraph_integer_t nodes,
                               igraph_integer_t types, igraph_integer_t k,
-                              igraph_vector_t *type_dist,
-                              igraph_matrix_t *pref_matrix,
-                              igraph_bool_t directed) {
-
+                              const igraph_vector_t *type_dist,
+                              const igraph_matrix_t *pref_matrix,
+                              igraph_bool_t directed,
+                              igraph_vector_t *node_type_vec) {
     long int i, j;
     igraph_vector_t edges;
     igraph_vector_t cumdist;
     igraph_vector_t potneis;
     igraph_real_t maxcum;
-    igraph_vector_t nodetypes;
+    igraph_vector_t *nodetypes;
+
+    /* Argument contracts */
+    if(nodes < 0){
+        IGRAPH_ERROR("The number of vertices must be non-negative.", IGRAPH_EINVAL);
+    }
+
+    if (types < 1) {
+        IGRAPH_ERROR("The number of vertex types must be at least 1.", IGRAPH_EINVAL);
+    }
+
+    if (type_dist) {
+        if (igraph_vector_size(type_dist) != types) {
+            IGRAPH_ERROR("The vertex type distribution vector must agree in length with the number of types.",
+                         IGRAPH_EINVAL);
+        }
+
+        if (igraph_vector_min(type_dist) < 0) {
+            IGRAPH_ERROR("The vertex type distribution vector must contain non-negative values.", IGRAPH_EINVAL);
+        }
+    }
+
+    if (igraph_matrix_nrow(pref_matrix) != types || igraph_matrix_ncol(pref_matrix) != types) {
+        IGRAPH_ERROR("The preference matrix must be square and agree in dimensions with the number of types.", IGRAPH_EINVAL);
+    }
+
+    if (! directed && ! igraph_matrix_is_symmetric(pref_matrix)) {
+        IGRAPH_ERROR("The preference matrix must be symmetric when generating undirected graphs.", IGRAPH_EINVAL);
+    }
+
+    {
+        igraph_real_t lo, hi;
+        igraph_matrix_minmax(pref_matrix, &lo, &hi);
+
+        if (lo < 0 || hi > 1) {
```

Comment:
What if a NaN somehow sneaks into the preference matrix? Both the comparisons will be `false` if I understand correctly and we won't hit this branch, right?

Reply:
OK, will update all of them to detect NaNs. I originally started this PR before we added NaN detection.

Reply:
Next we have to add some helpers which do the NaN check and error reporting automatically. Writing the same code again and again is getting tedious.

Reply:
A quick (although awkward) fix is to reverse the comparison: `!(lo >= 0 && hi <= 1)`

Reply:
Feel free to merge when you have updated the PR.